js/runner.go

```diff
@@ -356,65 +353,46 @@ func (r *Runner) timeoutErrorDuration(stage string) time.Duration {
 type VU struct {
 	BundleInstance
 
-	Runner    *Runner
-	Transport *http.Transport
-	Dialer    *netext.Dialer
-	CookieJar *cookiejar.Jar
-	TLSConfig *tls.Config
-	ID        int64
-	Iteration int64
+	Runner     *Runner
+	RunContext *context.Context
+	Transport  *http.Transport
+	Dialer     *netext.Dialer
+	CookieJar  *cookiejar.Jar
+	TLSConfig  *tls.Config
+	ID         int64
+	Iteration  int64
 
 	Console *console
 	BPool   *bpool.BufferPool
 
 	Samples chan<- stats.SampleContainer
 
 	setupData goja.Value
-
-	// A VU will track the last context it was called with for cancellation.
-	// Note that interruptTrackedCtx is the context that is currently being tracked, while
-	// interruptCancel cancels an unrelated context that terminates the tracking goroutine
-	// without triggering an interrupt (for if the context changes).
-	// There are cleaner ways of handling the interruption problem, but this is a hot path that
-	// needs to be called thousands of times per second, which rules out anything that spawns a
-	// goroutine per call.
-	interruptTrackedCtx context.Context
-	interruptCancel     context.CancelFunc
-
-	m *sync.Mutex
 }
 
-// Verify that VU implements lib.VU
-var _ lib.VU = &VU{}
+// Verify that interfaces are implemented
+var _ lib.ActiveVU = &VU{}
+var _ lib.InitializedVU = &VU{}
 
-func (u *VU) Reconfigure(id int64) error {
-	u.ID = id
-	u.Iteration = 0
-	u.Runtime.Set("__VU", u.ID)
-	return nil
-}
+// Activate the VU so it will be able to run code
+func (u *VU) Activate(params *lib.VUActivationParams) lib.ActiveVU {
+	u.Runtime.ClearInterrupt()
+	u.RunContext = &params.RunContext
+	// u.Env = params.Env
 
-func (u *VU) RunOnce(ctx context.Context) error {
-	u.m.Lock()
-	defer u.m.Unlock()
-	// Track the context and interrupt JS execution if it's cancelled.
-	if u.interruptTrackedCtx != ctx {
-		interCtx, interCancel := context.WithCancel(context.Background())
-		if u.interruptCancel != nil {
-			u.interruptCancel()
+	go func() {
```

Comment:
I tried spawning this goroutine conditionally, essentially moving the `DeactivateCallback != nil` check outside, but it didn't help with the test issues.

Reply:
I think it's correct to always spawn this goroutine.

Reply:
Yeah, after more consideration I agree with you. :) We need to interrupt the Goja `Runtime` in any case.

Initially I was concerned that this might spawn too many goroutines, but there should only be one per active VU, which doesn't seem excessive.